Assets/MixedRealityToolkit.Tests/PlayModeTests/InputEventSystemTests.cs

```diff
@@ -0,0 +1,300 @@
+        {
+            TestUtilities.InitializeMixedRealityToolkitAndCreateScenes(true);
+            TestUtilities.InitializePlayspace();
+
+            IMixedRealityInputSystem inputSystem = null;
+            MixedRealityServiceRegistry.TryGetService(out inputSystem);
+
+            Assert.IsNotNull(inputSystem, "Input system must be initialized");
+
+            // Let input system to register all cursors and managers.
+            yield return null;
+
+            // Switch off / Destroy all input components, which listen to global events
+            Object.Destroy(inputSystem.GazeProvider.GazeCursor as Behaviour);
+            inputSystem.GazeProvider.Enabled = false;
+
+            var diagnosticsVoiceControls = Object.FindObjectsOfType<DiagnosticsSystemVoiceControls>();
+            foreach (var diagnosticsComponent in diagnosticsVoiceControls)
+            {
+                diagnosticsComponent.enabled = false;
+            }
+
+            // Let objects be destroyed
+            yield return null;
+
+            // Check that input system is clean
+            Assert.AreEqual(((BaseEventSystem)inputSystem).EventHandlersByType.Count, 0, "Input event system handler registry is not empty in the beginning of the test.");
+
+            yield return null;
+        }
+
+        [TearDown]
+        public void ShutdownMrtk()
+        {
+            TestUtilities.ShutdownMixedRealityToolkit();
+        }
+
+        /// <summary>
+        /// </summary>
+        [UnityTest]
+        public IEnumerator Test01_TestObjectBasedEventRegistration()
+        {
+            yield return SetupMrtkWithoutCursors();
+
+            IMixedRealityInputSystem iInputSystem = null;
+            MixedRealityServiceRegistry.TryGetService(out iInputSystem);
+
+            BaseEventSystem inputSystem = (BaseEventSystem)iInputSystem;
+
+            var object1 = new GameObject("Object");
+
+            // Second handler weirdly depends on the first one due to event registration working on the entire object
+            var objectBasedListener = object1.AddComponent<TestInputGlobalListenerObjectBased>();
+            var handlerBasedListener = object1.AddComponent<TestInputGlobalListenerHandlerBasedSpeechHandler>();
+
+            yield return null;
+
+            // Event listener collection is filled for backward compatibility
+            Assert.AreEqual(inputSystem.EventListeners.Count, 1, "Event listener for old event system API hasn't been registered");
+            Assert.Contains(objectBasedListener.gameObject, inputSystem.EventListeners, "Wrong event listener registered for old event system API.");
+
+            Assert.AreEqual(inputSystem.EventHandlersByType.Count, 2, "Input event system doesn't contain expected event handler types.");
+            CollectionAssert.Contains(inputSystem.EventHandlersByType.Keys, typeof(IMixedRealitySpeechHandler), "Input event system doesn't contain IMixedRealitySpeechHandler entry.");
```

Comment:
When it comes to doing collection based comparisons, consider using something like:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.testtools.unittesting.collectionassert.areequivalent?view=mstest-net-1.2.0#Microsoft_VisualStudio_TestTools_UnitTesting_CollectionAssert_AreEquivalent_System_Collections_ICollection_System_Collections_ICollection_

This prevents you from having to do a size assert first, followed by an individual assert for each item. 

The only reason you shouldn't do this is if that AreEquivalent actually gives you a really bad diff/error message.

The reason I also ask is, the way it's written, if something is missing (i.e. count is 1) you'll only get an error that the count is 1, and you'll never execute the other stuff which tells you specifically what is missing.

tl;dr when doing collection based comparisons, used collection based compare functions, and if you want to do individual ones prefer doing the size check at the end to get the most helpful error messages.

Reply:
CollectionAssert comparisons are awesome!